### PR TITLE
Refactor of the big if-chain to a dictionary in the form {backend_name: backend_open}.

### PR DESCRIPTION
**Pull request by alexamici**
This ports the clean up of the messy if-else-chain from #3166. It should make adding the new back-end API easier.

The change is relatively trivial, but it touches code in `xr.open_dataset` so extra care should be used by reviewers. We think the change is safe and it does pass all tests.

 - [x] Port the generic part of #3166
 - [x] Tests sync'ed
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] No user visible change
 - [x] No new functions/methods

I cannot add reviewers myself, but we agreed this should go to @jhamman.

### CODE REVIEW

**Comment by alexamici** [comment]
@jhamman Stephan requested to keep this PR on hold until #4187 is merged as it will produce conflicts.

So I converted the PR to draft, after #4187 gets in, we will fix the conflicts and resubmit.

**Comment by jhamman** [comment]
Thanks for the ping @alexamici - this looks good to me. Happy to see this merged once we sort out the conflicts in #4187.

**Comment by shoyer** [comment]
https://github.com/pydata/xarray/pull/4187 has been merged, so this can go ahead again!

**Comment by alexamici** [comment]
Woops I didn't know that force pushing a branch would close the PR :/

I'll try to reopen this PR or create a new one. Sorry for the noise.

**Comment by dopplershift** [comment]
@alexamici Force-pushing doesn't normally close it, so this is weird...

**Comment by keewis** [comment]
the reason the PR was closed is that you removed the changes, so the diff to `master` is empty. I think you can reopen after pushing commits with changes.

**Comment by alexamici** [comment]
Thanks @keewis! So it is correct, I wanted to reset to master and re apply the changes.

**Comment by alexamici** [comment]
@jhamman I don't understand why 4f414f2 is shown here, it is a commit in master.

As for the rest, we re applied the engine selection refactor and added the new `zarr` bits.

All looks reasonable to me and all tests pass.

**Comment by shoyer** [comment]
> @jhamman I don't understand why [4f414f2](https://github.com/pydata/xarray/commit/4f414f2d5eb2e5a12fb8ae1012c5ac7aa43b6f0b) is shown here, it is a commit in master.

I usually find it easiest just to use "merge master" to integrate the latest changes, and to treat pull requests as "append only". History gets consolidated/rewritten when we merge the PR with "Squash and Merge"

**Review by shoyer — APPROVED** [review]

File: xarray/backends/api.py
Comment (jhamman) [inline xarray/backends/api.py:547]: I think this was a bug introduced in #4187. Not exactly sure what we want but we used to support zarr filenames as strings. This line now requires they be a `MutableMapping`. Curious to hear from others but we may be able to get away with the following.

```suggestion
        elif engine == "zarr":
```

Comment (jhamman) [reply xarray/backends/api.py:547]: cc @weiji14 on this one!

Comment (weiji14) [reply xarray/backends/api.py:547]: See https://github.com/intake/intake-xarray/issues/70#issuecomment-621268486 for context, if I'm reading it correctly, it's so that remote Zarr files can be accessed via [`fsspec.get_mapper`](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.get_mapper).

Comment (jhamman) [reply xarray/backends/api.py:547]: @weiji14 - The problem with the current implementation is that `open_dataset("./filename.zarr", engine="zarr")` will raise the `ValueError("can only read bytes...")` before getting down to lines 555:557.

To illustrate further, we caught this in our nightly upstream test on xpublish: https://github.com/xarray-contrib/xpublish/runs/1156500330?check_suite_focus=true

Comment (alexamici) [reply xarray/backends/api.py:547]: @jhamman the `if isinstance(filename_or_obj, str):` at line 542 handles the `str` case for 'zarr' before the `elif`. No?

Comment (alexamici) [reply xarray/backends/api.py:547]: I confirm, in this PR paths passed as strings reach line 573 as expected:

`store = opener(filename_or_obj, **extra_kwargs, **backend_kwargs)`

Comment (keewis) [reply xarray/backends/api.py:547]: @jhamman, if I'm reading the logs correctly, `filename_or_obj` is a `tests.utils.TestMapper` object which inherits from `starlette.testclient.TestClient`. I think the tests might pass again if we use `xarray.core.utils.is_dict_like(...)` instead of `isinstance(..., MutableMapping)`

Comment (shoyer) [reply xarray/backends/api.py:547]: Can we avoid all type checks on `filename_or_obj` when engine is explicitly provided? It seems like we could safely just pass it on to the datastore object?

Comment (keewis) [reply xarray/backends/api.py:547]: I just realized we've been talking about the changes in current `master`, and most of this is already fixed by this PR. If it makes sense to always set `mode="r"` I guess we can simply remove the `isinstance` check.

Comment (alexamici) [reply xarray/backends/api.py:547]: @shoyer 

> Can we avoid all type checks on `filename_or_obj` when engine is explicitly provided? It seems like we could safely just pass it on to the datastore object?

Not sure. Most type checks conditionals apply generic transformations, like conversion from `Path` to `str` and `_normalize_path` that are probably needed by all backends.

Also, the logic for the selection of the default engine varies depending on the `filename_or_obj` data type.

I only see a possible simplification in the handling of the opened-file case but at the cost of changing the current behaviour (in `master` and in this PR if you attempt to open a netCDF3 with `engine='h5netcdf'` you end up silently using `engine='scipy'`).

I would suggest to make a different PR for that change.

Comment (alexamici) [reply xarray/backends/api.py:547]: I can add this change that makes `mode="r"` unconditional for 'zarr' and cleans up a bit the ifs (Note sure how to show you without making the commit):

```diff
Index: xarray/backends/api.py
I===================================================================
--- xarray/backends/api.py	(revision 4a1319243692e2a4425f9a7faf0c5df8a80a630d)
+++ xarray/backends/api.py	(date 1600901329932)
@@ -544,12 +543,7 @@
 
             if engine is None:
                 engine = _get_default_engine(filename_or_obj, allow_remote=True)
-        elif isinstance(filename_or_obj, MutableMapping) and engine == "zarr":
-            # Zarr supports a wide range of access modes, but for now xarray either
-            # reads or writes from a store, never both.
-            # For open_dataset(engine="zarr"), we only read (i.e. mode="r")
-            extra_kwargs["mode"] = "r"
-        else:
+        elif engine != "zarr":
             if engine not in [None, "scipy", "h5netcdf"]:
                 raise ValueError(
                     "can only read bytes or file-like objects "
@@ -567,6 +561,7 @@
             overwrite_encoded_chunks = backend_kwargs.pop(
                 "overwrite_encoded_chunks", None
             )
+            extra_kwargs["mode"] = "r"
             extra_kwargs["group"] = group
 
         opener = _get_backend_cls(engine)
```

Comment (alexamici) [reply xarray/backends/api.py]: > I can add this change...

Done

**Review by jhamman — APPROVED** [review]

**Pull request merged**
